### ISSUE TITLE
[Differential evolution:] Support for parallel evaluations of the constraint function.

### ISSUE THREAD

**Issue by pzuvela** [scipy/scipy#12628]
**Is your feature request related to a problem? Please describe.**
Currently working on a high-dimensional constrained optimization problem (typical matrix size: ~12,000 × 5,000) with cca. 30-40 parameters to optimize over. Typical number of iterations is 10, and population size of 1 (keyword arg value). Such a size is sufficient because the optimization surface is quite flat. Unfortunately, I am under an NDA and cannot reveal too many details about the problem.

**Describe the solution you'd like**
In the current implementation, only the objective function evaluations are carried out in parallel (i.e., within **"_calculate_population_energies"** there is a call to the MapWrapper object instantiated in the constructor which evaluates the _i_-th population in parallel). However, for constrained optimization, evaluations of the constraint function are carried out in serial mode. The **"_calculate_population_feasibilities"** method is not parallelized. In the case of my problem, one evaluation of the constraint function is expensive and can last for 4-5 or more mins. Therefore, parallelization would yield computational cost savings.

**Describe alternatives you've considered**
N/A

**Additional context (e.g. screenshots)**
I have written an in-house implementation by deriving from the DifferentialEvolutionSolver class and over-riding the **"_calculate_population_feasibilities"** method. Using the joblib library (Parallel, and delayed), the constraint violation values are calculated in parallel via the following notation:
```
        constraint_violation = np.array(list(Parallel(n_jobs=self.constraint_workers, verbose=10)(
            delayed(np.concatenate)([c.violation(params) for c in self._wrapped_constraints])
            for params in parameters_pop)))
```
Such a work-around, instead of a direct call to the **"constraint_violation_fn"** method is due to the same not being picklable. To employ the multiprocessing package from the standard library would be also possible.
In order to prevent over-subscribing resources, I added a new keyword to the constructor entitled **"constraint_workers"**.
If such a feature would yield significant improvements (still due to be tested), the implementation would be of potential use to other SciPy DE users.

**Comment by andyfaff** [comment]
This seems like a good idea. Is it worthbthe addition of the extra keyword, or would parallelising the constraint functions always occur alongside parallelising evaluation of the main objective? Your constraint function sounds enormous, I've only ever experienced cheap constraint functions.

**Comment by pzuvela** [comment]
If the resources are released (i.e., the processing pool is not being recycled for each generation), then we may not require a new keyword, as the resources will not get over-subscribed. On the other hand, if in the usual case the constraint function is cheap, then the overhead of including many workers would "eat up" the benefits. Then, a separate keyword for the number of constraint function workers is worthwhile.

Behind both the objective and constraint functions in my problem, there is a machine learning (ML) model underneath which needs to be generated to make predictions. Several constraints are evaluated from such an ML model. To decrease computational costs, I've linked the optimization and constraint functions via a flag as an input argument to the objective function (e.g., constr=True). Objective function is always evaluated, and if the flag is set to True it has an additional call to the constraint function. Re-factoring this may also lead to a more effective solution?

In any case, I will generalize my approach to use the native multi-processing module, and begin a pull request.

**Comment by andyfaff** [comment]
The `MapWrapper` is used for the duration of the minimisation, so it can be reused cheaply.
If your model is expensive to calculate then you may be able to perform some cacheing. The constraint function is called for all the population members, then the objective function is calculated for all the feasible entries. If your constraint function calculates the objective as part of it's calculation, then you could cache the objective value (for a given `x`) to save time.

The difficulties with reusing `workers` for both objective and constraint are:

- specifying `workers` means that `updating` has to be `deferred`.
- what happens if one or other of the objective/constraints functions are not pickleable?

**Comment by pzuvela** [comment]
Thank you for the suggestion. I will explore caching to save time.

MapWrapper can still be re-used. However, what would happen in a case that the objective function is expensive, and the constraint function is notably less expensive? Let the number of workers at the same value for both evaluation calls (e.g., 8). The parallel processing overheads may be fine for the expensive objective function. But, they could actually lead to a slowdown for the constraint function. Currently, are there any mechanisms to reduce the number of workers not to over-subscribe resources? **For instance, if the population size is 10, and the user sets the number of workers to 32?**


1) Requirement to set the updating to "deferred" is a good reason for having the same keyword in both cases.

2) In the case of my problem, the constraint function is static and does not have elements that cannot be pickled. Being called from the objective function, since the objective function is pickleable, so is the constraint function. To generalize it, either a failure will occur (an exception/error) when both functions are not pickleable, or parallelization can be disabled for the constraint function by default if the function is not pickleable.

**Comment by andyfaff** [comment]
>  Currently, are there any mechanisms to reduce the number of workers not to over-subscribe resources? For instance, if the population size is 10, and the user sets the number of workers to 32?

That wouldn't make sense for the user to do that. The action taken by the default MapWrapper would be whatever Python itself does. Setting workers=-1 uses the maximum number of CPU.

If the objective function is really expensive then the amount of overhead added by evaluating the constraint function in parallel will probably be minimal in comparison.

**Comment by pzuvela** [comment]
Okay then.

Is it safe to say the potential new feature would have:
1) Shared `workers `keyword for evaluating both objective and constraint function in parallel (no changes to `deferred`/`immediate` mechanism necessary).
2) Exception if constraint function is not pickleable.
3) Alternative to 2) to disable parallel processing if the constraint function is pickleable.

**Comment by andyfaff** [comment]
for (3) I would `try/except PicklingError`, falling back to `map`.

**Comment by pzuvela** [comment]
Perfect. Thank you for your comments. I will prepare an update to the DE code and attach a PR to the thread once ready. 

@andyfaff  : since this is an addition to the existing DE feature in SciPy, do you think it is required to update the unit tests for DE? Or the current coverage is sufficient... ?

**Comment by andyfaff** [comment]
Additional tests will be required, that examine whether the constraints can be called in serial or parallel fashion, and whether the fallback to serial works.

**Comment by andyfaff** [comment]
The code changes will be along the lines of:

```
   try:
        constraint_violation = np.array(list(self._mapwrapper(self._constraint_violation_fn(x), parameters_pop)))
    except (TypeError, ValueError) as e:
            # wrong number of arguments for _mapwrapper
            # or wrong length returned from the mapper
            raise RuntimeError(
                "The map-like callable must be of the form f(func, iterable), "
                "returning a sequence of numbers the same length as 'iterable'"
            ) from e
    except PicklingError as e:
        # deal with constraints that aren't pickleable
        warnings.warn("The constraint functions must be pickleable for parallel evaluation, falling back to map", UserWarning)
        constraint_violation = np.array(list(map(self._constraint_violation_fn(x), parameters_pop)))
    feasible = ~(np.sum(constraint_violation, axis=1) > 0)
```

**Comment by pzuvela** [comment]
Thanks ! 
I tested this and after correcting the syntax error it works like a charm.

The argument `x` to `_constraint_violation_fn` method would be undefined here. Simply remove `(x)` from both mappers.

In a list comprehension it would be required:
e.g., `[self._constraint_violation_fn(x) for x in parameters_pop]`

What would be the next step for potentially including this new feature into the DE module?

The code below is within my project (deriving from the DifferentialEvolutionSolver class; in a custom de.py module).
In the same de.py custom module in my project is another `de` function which just instantiates DifferentialEvolution class instead of DifferentialEvolutionSolver class.

```
class DifferentialEvolution(DifferentialEvolutionSolver):

    # Over-ride the constructor
    def __init__(self, func, bounds, args=(), strategy='best1bin', maxiter=1000, popsize=15,
                 tol=0.01, mutation=(0.5, 1), recombination=0.7, seed=None,
                 maxfun=np.inf, callback=None, disp=False, polish=True,
                 init='latinhypercube', atol=0, updating='immediate', workers=1, constraints=(),
                 constraint_workers=1):

        super().__init__(func, bounds, args, strategy, maxiter, popsize, tol,
                         mutation, recombination, seed, maxfun, callback, disp, polish,
                         init, atol, updating, workers, constraints)

        self.workers = workers

        """def __del__(self):
            pass"""

        """def __exit__(self, *args):
            pass"""

        # Number of constraint workers
        self.constraint_workers = constraint_workers

    def _calculate_population_feasibilities(self, population):

        """
        Calculate the feasibilities of a population.

        Parameters
        ----------
        population : ndarray
            An array of parameter vectors normalised to [0, 1] using lower
            and upper limits. Has shape ``(np.size(population, 0), len(x))``.

        Returns
        -------
        feasible, constraint_violation : ndarray, ndarray
            Boolean array of feasibility for each population member, and an
            array of the constraint violation for each population member.
            constraint_violation has shape ``(np.size(population, 0), M)``,
            where M is the number of constraints.
        """
        num_members = np.size(population, 0)
        if not self._wrapped_constraints:
            # shortcut for no constraints
            return np.ones(num_members, bool), np.zeros((num_members, 1))

        parameters_pop = self._scale_parameters(population)

        if self.workers > num_members:
            self.workers = num_members
            warnings.warn("Number of workers larger than the number of members, "
                          "decreasing parallel processing pool....", UserWarning)

        try:
            constraint_violation = np.array(list(self._mapwrapper(self._constraint_violation_fn, parameters_pop)))
        except (TypeError, ValueError) as e:
            # wrong number of arguments for _mapwrapper
            # or wrong length returned from the mapper
            raise RuntimeError(
                "The map-like callable must be of the form f(func, iterable), "
                "returning a sequence of numbers the same length as 'iterable'"
            ) from e
        except PicklingError as e:
            # deal with constraints that aren't pickleable
            warnings.warn("The constraint functions must be pickleable for parallel evaluation, falling back to map",
                          UserWarning)
            constraint_violation = np.array(list(map(self._constraint_violation_fn, parameters_pop)))
        feasible = ~(np.sum(constraint_violation, axis=1) > 0)

        return feasible, constraint_violation
```

**Comment by pzuvela** [comment]
Thanks !
Your code had a syntax error. The argument x to _constraint_violation_fn method would be undefined here. Simply remove (x) from both mappers. In a list comprehension it would be required: e.g., [self._constraint_violation_fn(x) for x in parameters_pop].
I tested this, and a PicklingError is raised: **`pool objects cannot be passed between processes or pickled`**.

The code below is within my project (deriving from the DifferentialEvolutionSolver class; in a custom de.py module).
In the same de.py custom module in my project is another de function which just instantiates DifferentialEvolution class instead of DifferentialEvolutionSolver class. If the `joblib` package is employed, a list comprehension can be used instead of a mapping. Instead of calling the private method `_constraint_violation_fn`, a call can be made to `concatenate` (from numpy) instead. In this case, no pickling errors are raised and the code runs correctly. Any comments on how to refactor that into a mapper?

```
class DifferentialEvolution(DifferentialEvolutionSolver):

    # Over-ride the constructor
    def __init__(self, func, bounds, args=(), strategy='best1bin', maxiter=1000, popsize=15,
                 tol=0.01, mutation=(0.5, 1), recombination=0.7, seed=None,
                 maxfun=np.inf, callback=None, disp=False, polish=True,
                 init='latinhypercube', atol=0, updating='immediate', workers=1, constraints=()):

        super().__init__(func, bounds, args, strategy, maxiter, popsize, tol,
                         mutation, recombination, seed, maxfun, callback, disp, polish,
                         init, atol, updating, workers, constraints)

        self.workers = workers

    def _calculate_population_feasibilities(self, population):

        """
        Calculate the feasibilities of a population.

        Parameters
        ----------
        population : ndarray
            An array of parameter vectors normalised to [0, 1] using lower
            and upper limits. Has shape ``(np.size(population, 0), len(x))``.

        Returns
        -------
        feasible, constraint_violation : ndarray, ndarray
            Boolean array of feasibility for each population member, and an
            array of the constraint violation for each population member.
            constraint_violation has shape ``(np.size(population, 0), M)``,
            where M is the number of constraints.
        """
        num_members = np.size(population, 0)
        if not self._wrapped_constraints:
            # shortcut for no constraints
            return np.ones(num_members, bool), np.zeros((num_members, 1))

        parameters_pop = self._scale_parameters(population)
        try:
            **constraint_violation = np.array(list(Parallel(n_jobs=self.workers, verbose=10)(
                delayed(np.concatenate)([c.violation(params) for c in self._wrapped_constraints])
                for params in parameters_pop)))**
        except (TypeError, ValueError) as e:
            # wrong number of arguments for _mapwrapper
            # or wrong length returned from the mapper
            raise RuntimeError(
                "The map-like callable must be of the form f(func, iterable), "
                "returning a sequence of numbers the same length as 'iterable'"
            ) from e

        except PicklingError as e:
            # deal with constraints that aren't pickleable
            warnings.warn("The constraint functions must be pickleable for parallel evaluation, falling back to map",
                          UserWarning)
            constraint_violation = np.array(list(map(self._constraint_violation_fn(x), parameters_pop)))

        feasible = ~(np.sum(constraint_violation, axis=1) > 0)

        return feasible, constraint_violation
```

**Comment by andyfaff** [comment]
The mapwrapper call should look something like the calculate population energies call.

**Comment by andyfaff** [comment]
#15439 adds the `vectorized` keyword (available in 1.9.x onwards). This sends several solution vectors (typically the entire population) to the objective *and* the constraint functions, with those functions returning arrays corresponding to the outputs of the objective and constraints.
This would allow the user to implement their own parallelisation inside either the objective or constraint functions. The following example shows a user implemented parallelisation inside the objective function

```
from scipy.optimize import differential_evolution, rosen
from multiprocessing import Pool

bounds = [(0, 10)] * 2
with Pool(4) as p:
    def fun(x):
        # x.shape = (N, S) where N is the number of parameters and S is the
        # number of solution vectors to be considered. So we need to transpose it
        # if we're going to send it to a map-like
        if x.ndim == 1:
            # when polishing is performed
            return rosen(x)
        else:
            return np.array(list(p.map(rosen, x.T)))
    
    res = differential_evolution(fun, bounds, vectorized=True)
```

It should be possible to parallelise constraints in a similar manner.
